planning/behavior_path_planner/src/scene_module/avoidance/avoidance_utils.cpp

```diff
@@ -342,4 +342,149 @@ Polygon2d createEnvelopePolygon(
+  if (points.size() <= 2) {
+    // sort data based on longitudinal distance to the boundary
+    std::sort(
+      sorted_points.begin(), sorted_points.end(),
+      [](const PolygonPoint & a, const PolygonPoint & b) { return a.lon_dist < b.lon_dist; });
+    return sorted_points;
+  }
+
+  // sort data based on lateral distance to the boundary
+  std::sort(
+    sorted_points.begin(), sorted_points.end(), [](const PolygonPoint & a, const PolygonPoint & b) {
+      return std::fabs(a.lat_dist_to_bound) > std::fabs(b.lat_dist_to_bound);
+    });
+  PolygonPoint first_point;
+  PolygonPoint second_point;
+  if (sorted_points.at(0).lon_dist < sorted_points.at(1).lon_dist) {
+    first_point = sorted_points.at(0);
+    second_point = sorted_points.at(1);
+  } else {
+    first_point = sorted_points.at(1);
+    second_point = sorted_points.at(0);
+  }
+
+  for (size_t i = 2; i < sorted_points.size(); ++i) {
+    const auto & next_point = sorted_points.at(i);
+    if (next_point.lon_dist < first_point.lon_dist) {
+      return {next_point, first_point, second_point};
+    } else if (second_point.lon_dist < next_point.lon_dist) {
+      return {first_point, second_point, next_point};
+    }
+  }
+
+  return {first_point, second_point};
+}
+
+std::vector<PolygonPoint> getEdgePoints(
+  const std::vector<Point> & bound, const std::vector<Point> & edge_points,
+  const double lat_dist_to_path, size_t & start_segment_idx, size_t & end_segment_idx)
+{
+  std::vector<PolygonPoint> edge_points_data;
+  for (const auto & edge_point : edge_points) {
+    const size_t segment_idx = motion_utils::findNearestSegmentIndex(bound, edge_point);
+    start_segment_idx = std::min(start_segment_idx, segment_idx);
+    end_segment_idx = std::max(end_segment_idx, segment_idx);
+
+    PolygonPoint edge_point_data;
+    edge_point_data.point = edge_point;
+    edge_point_data.lat_dist_to_bound = motion_utils::calcLateralOffset(bound, edge_point);
+    edge_point_data.lon_dist = motion_utils::calcSignedArcLength(bound, 0, edge_point);
+    if (lat_dist_to_path >= 0.0 && edge_point_data.lat_dist_to_bound > 0.0) {
+      continue;
+    } else if (lat_dist_to_path < 0.0 && edge_point_data.lat_dist_to_bound < 0.0) {
+      continue;
+    }
+
+    edge_points_data.push_back(edge_point_data);
+  }
+
+  return edge_points_data;
+}
+
+std::vector<Point> updateBoundary(
+  const std::vector<Point> & original_bound, const std::vector<PolygonPoint> & points,
+  const size_t start_segment_idx, const size_t end_segment_idx)
+{
+  if (start_segment_idx >= end_segment_idx) {
+    return original_bound;
+  }
+
+  std::vector<Point> updated_bound;
+  for (size_t i = 0; i <= start_segment_idx; ++i) {
+    updated_bound.push_back(original_bound.at(i));
+  }
+  for (size_t i = 0; i < points.size(); ++i) {
+    updated_bound.push_back(points.at(i).point);
+  }
+  for (size_t i = end_segment_idx + 1; i < original_bound.size(); ++i) {
+    updated_bound.push_back(original_bound.at(i));
+  }
```

Comment:
It maybe use simpler to use std::copy.
reference : https://www.paveway.info/entry/2019/10/09/cpp_copyvector

Reply:
There's still one :pray: 
![image](https://user-images.githubusercontent.com/8327598/206487090-dd67eeda-326d-450e-b335-da16d99bdb88.png)

Reply:
The middle cannot be replace by std::copy since it get points from `points.at(I).point`

Reply:
I see :+1: